scipy/optimize/_zeros_py.py

```diff
@@ -1392,3 +1393,344 @@ def toms748(f, a, b, args=(), k=1,
+    by ``func``.
+
+    References
+    ----------
+
+    .. [1] Chandrupatla, Tirupathi R.
+        "A new hybrid quadratic/bisection algorithm for finding the zero of a
+        nonlinear function without using derivatives".
+        Advances in Engineering Software, 28(3), 145-149.
+        https://doi.org/10.1016/s0965-9978(96)00051-8
+
+    See Also
+    --------
+    brentq, brenth, ridder, bisect, newton
+
+    Examples
+    --------
+    >>> from scipy import optimize
+    >>> def f(x, c):
+    ...     return x**3 - 2*x - c
+    >>> c = 5
+    >>> res = optimize._zeros_py._chandrupatla(f, 0, 3, args=(c,))
+    >>> res.x
+    2.0945514818937463
+
+    >>> c = [3, 4, 5]
+    >>> res = optimize._zeros_py._chandrupatla(f, 0, 3, args=(c,))
+    >>> res.x
+    array([1.8932892 , 2.        , 2.09455148])
+
+    """
+
+    res = _chandrupatla_iv(func, a, b, args, xatol, xrtol,
+                           fatol, frtol, maxiter, callback)
+    func, a, b, args, xatol, xrtol, fatol, frtol, maxiter, callback = res
+
+    # Initialization
+    x1, f1, x2, f2, shape, dtype = _initialize_xf(func, a, b, args)
+    status = np.full_like(x1, _EINPROGRESS, dtype=int)  # in progress
+    active = np.arange(x1.size)  # indices of in-progress elements
+    nit, nfev = 0, 2  # two function evaluations performed above
+    cb_terminate = False
+    fatol = fatol or np.finfo(dtype).smallest_normal
+    frtol = frtol * np.minimum(np.abs(f1), np.abs(f2))
+    tols = dict(xatol=xatol, xrtol=xrtol, fatol=fatol, frtol=frtol)
+    # Elements of `x1`, `f1`, etc., are stored in this `OptimizeResult`
+    # once a termination condition is met, and then the arrays are condensed
+    # to reduce unnecessary computation.
+    res = OptimizeResult(x=x1.copy(), fun=f1.copy(), xl=x1.copy(),
+                         fl=f1.copy(), xr=x2.copy(), fr=f2.copy(),
+                         nit=np.full_like(status, nit)[()],
+                         nfev=np.full_like(status, nfev)[()],
+                         status=status.copy(), success=(status==0))
+
+    temp = _check_termination(x1, f1, x2, f2, None, None,
+                              res, active, status, nfev, nit, tols)
+    xmin, fmin, x1, f1, x2, f2, x3, f3, active, status, tl = temp
+
+    if callback is not None:
+        temp_res = _prepare_result(shape, res, active, nit, nfev)
+        if _call_callback_maybe_halt(callback, temp_res):
+            cb_terminate == True
+
+    t = 0.5
+    while nit < maxiter and active.size and not cb_terminate:
+        # [1] Flowchart 1
+        x = x1 + t * (x2 - x1)
+
+        # For now, we assume that `func` requires arguments of the original
+        # shapes. However, `x` is compressed (contains only active elements).
+        # Therefore, we inflate `x` by creating a full-size array, copying the
+        # elements of `x` into it, and making it the expected shape.
+        # TODO: allow user to specify that `func` works with compressed input
```

Comment:
This will be in a follow-up. It will make the function much faster because then we won't have to "inflate" the argument, evaluate `func` with all elements (many of which may be unnecessary because they won't be used), and re-compress the output.

Reply:
Thanks for discussing this. I'll take the liberty of resolving this one.